vnext/ReactUWP/Views/Image/ReactImage.cpp

```diff
@@ -52,66 +51,194 @@ void ReactImage::OnLoadEnd(winrt::event_token const &token) noexcept {
+  if (source.uri.length() == 0) {
     m_onLoadEndEvent(*this, false);
     return;
   }
 
-  if (source.packagerAsset && uriString.find("file://") == 0) {
-    uriString.replace(0, 7, source.bundleRootPath);
-  }
+  m_imageSource = source;
 
-  winrt::Uri uri{Utf8ToUtf16(uriString)};
+  winrt::Uri uri{Utf8ToUtf16(m_imageSource.uri)};
   winrt::hstring scheme{uri.SchemeName()};
-  bool needsDownload = (scheme == L"http") || (scheme == L"https");
-  bool inlineData = scheme == L"data";
+
+  if (((scheme == L"http") || (scheme == L"https")) && !m_imageSource.headers.empty()) {
+    m_imageSource.sourceType = ImageSourceType::Download;
+  } else if (scheme == L"data") {
+    m_imageSource.sourceType = ImageSourceType::InlineData;
+  }
+
+  SetBackground(true);
+}
+
+winrt::IAsyncOperation<winrt::InMemoryRandomAccessStream> ReactImage::GetImageMemoryStreamAsync(
+    ReactImageSource source) {
+  switch (source.sourceType) {
+    case ImageSourceType::Download:
+      return co_await GetImageStreamAsync(source);
+    case ImageSourceType::InlineData:
+      return co_await GetImageInlineDataAsync(source);
+    default: // ImageSourceType::Uri
+      return nullptr;
+  }
+}
+
+winrt::fire_and_forget ReactImage::SetBackground(bool fireLoadEndEvent) {
+  const ReactImageSource source{m_imageSource};
+  const winrt::Uri uri{Utf8ToUtf16(source.uri)};
+  const bool fromStream{source.sourceType != ImageSourceType::Uri};
+
+  winrt::InMemoryRandomAccessStream memoryStream{nullptr};
+
   // get weak reference before any co_await calls
   auto weak_this{get_weak()};
 
   try {
-    m_imageSource = source;
+    memoryStream = co_await GetImageMemoryStreamAsync(source);
 
-    winrt::InMemoryRandomAccessStream memoryStream;
-    if (needsDownload) {
-      memoryStream = co_await GetImageStreamAsync(source);
-    } else if (inlineData) {
-      memoryStream = co_await GetImageInlineDataAsync(source);
-    }
-
-    if (auto strong_this{weak_this.get()}) {
-      if ((needsDownload || inlineData) && !memoryStream) {
+    // Fire failed load event if we're not loading from URI and the memory stream is null.
+    if (fromStream && !memoryStream) {
+      if (auto strong_this{weak_this.get()}) {
         strong_this->m_onLoadEndEvent(*strong_this, false);
       }
+      return;
+    }
+  } catch (winrt::hresult_error const &) {
+    const auto strong_this{weak_this.get()};
+    if (strong_this && fireLoadEndEvent) {
+      strong_this->m_onLoadEndEvent(*strong_this, false);
+    }
+  }
 
-      if (!needsDownload || memoryStream) {
-        auto surface = needsDownload || inlineData ? winrt::LoadedImageSurface::StartLoadFromStream(memoryStream)
-                                                   : winrt::LoadedImageSurface::StartLoadFromUri(uri);
+  if (auto strong_this{weak_this.get()}) {
+    if (strong_this->m_useCompositionBrush) {
+      const auto compositionBrush{ReactImageBrush::Create()};
+      compositionBrush->AvailableSize(strong_this->ActualSize());
```

Comment:
>ActualSize [](start = 51, length = 10)

ActualSize won't get set/updated until after XAML layout runs on a later tick.  I don't see any guarantee that will happen before you access ActualSize.

Reply:
There should be valid after [LayoutUpdated](https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.frameworkelement.layoutupdated) although [SizeChanged](https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.frameworkelement.sizechanged) is probably more appropriate. Is it critical that we catch this mid-layout rather than after? Unless so it seems more appropriate to wait for the standard events.

Reply:
Changed it to use ReactImage's SizeChanged to set the initial available size of ReactImageBrush. The handler is then revoked in LIS.LoadCompleted when the ReactImageBrush is set to the ReactImage's background. 

We're still using ActualSize when we're dynamically switching the resizeMode to repeat though. That's because in that case, both SizeChanged and LayoutUpdate have already been fired.